proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,343 @@
+  - `u32` Byte offset at which to read
+  - `u32` Length of the chunk to copy in bytes
+- Behavior:
+  - Check there are at least three instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and source account do not alias,
+  otherwise throw `AccountBorrowFailed`
+  - Verify the program account
+  - Check the status stored in the program account is retracted,
+  otherwise throw `InvalidArgument`
+  - Check that the source account is owned by loader v1, v2, v3 or v4,
+  otherwise throw `InvalidArgument`
+  - and look-up the source header size:
+    - loader-v1: 0 bytes
+    - loader-v2: 0 bytes
+    - loader-v3: 45 bytes
+    - loader-v4: 48 bytes
+  - Check that the source end offset (sum of source offset and length) does
+  not exceed the maximum (source account length minus the source header size),
+  otherwise throw `AccountDataTooSmall`
+  - Check that the destination end offset (sum of destination offset and
+  length) does not exceed the maximum (program account length minus the loader-v4
+  header size), otherwise throw `AccountDataTooSmall`
+  - Copy the chunk between the program accounts at the offsets, each shifted by
+  the header size of their loader (account owner) respectively
+
+#### SetProgramLength
+
+- Instruction accounts:
+  - `[writable]` The program account to change the size of.
+  - `[signer]` The authority of the program.
+  - `[writable]` Optional, the recipient account.
+- Instruction data:
+  - Enum variant `2u32`
+  - `u32` The new size after the operation.
+- Behavior:
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - If this is an initialization (program account length is too short to
+  contain the header):
+    - the owner of the program account is loader-v4,
+    otherwise throw `InvalidAccountOwner`
+    - the program account is writable, otherwise throw `InvalidArgument`
+    - the provided authority (instruction account at index 1) signed,
+    otherwise throw `MissingRequiredSignature`
+  - If this is not an initialization:
+    - Verify the program account
+    - Check that the status stored in the program account is retracted,
+    otherwise throw `InvalidArgument`
+  - Check that there are enough funds in the program account for rent
+  exemption, otherwise throw `InsufficientFunds`
+  - If there are more than enough funds:
+    - Check there are at least three instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+    - Check that the recipient account (instruction account at index 2) is
+    writable, otherwise throw `InvalidArgument`
+    - If a recipient account was provided that is not the program account:
+      - Transfer the surplus from the program account to the recipient account
+    - otherwise, if the requested new size is zero throw `InvalidArgument`
+  - If the requested new size is zero:
+    - Delete the entire program account, including the header
```

Comment:
Even if the program account is deleted, its `is_executable` flag will still be set to `true` and another instruction could transfer lamports to the account to keep it alive. This would create a weird edge case where a loader-v4 owned account would have its `is_executable` flag set to `true` despite not having a program state.

Reply:
We don't assign the account to the system program, so the actual deletion only happens implicitly at the end of the transaction. In between, no other instruction in the transaction which is not running loader-v4 can do anything with the account.

Reply:
Yeah, my point is that a _later_ instruction in the transaction could top up the balance of the closed program account such that it doesn't get deleted at the end of the transaction.

Reply:
Then it would stay a loader-v4 program. So the only thing you could get is an uninitialized loader-v4 program which has the executable flag set. However, you can already have a retracted loader-v4 program which has the executable flag the "official" way, and that is quite similar.

Reply:
Yes, exactly. This is what I said is the issue:
> This would create a weird edge case where a loader-v4 owned account would have its is_executable flag set to true despite not having a program state

Reply:
The plan is to emulate the `is_executable` flag in the VM using only the owner field, regardless of the program state. So this is already consistent with that goal.

Reply:
Sorry I don't really understand your response

Reply:
As long as `is_executable` is set and readable in the VM, I think we should go with because it's the most "correct" and not as complex as the verified option you mentioned.
> Make it true depending on the loader state: This is more complex and depends on the loaders.

If we go ahead with further cleanup of the `is_executable` flag in a new SIMD, I think we should do that first and then this SIMD doesn't need to worry about the flag at all.

Reply:
Then this SIMD would have to specify that:

- The flag is set when the program is deployed
- The flag is reset when the program is retracted

Reply:
Hmm maybe? I don't really understand why loader-v4 needs the ability to "retract" programs in addition to closing them.

Reply:
The retract state is what tells the program runtime that a program needs to be re-verified before it can be deployed. That way we can allow multiple modifications before one shared verification. Also, it allows a modular composition of these modifications instead of requiring a special instruction for each use-case / combination of those.